### PR TITLE
fix: support pending learner/admin users tied to more than 1 Enterprise

### PR DESCRIPTION
**Pull request by adamstankiewicz**
https://openedx.atlassian.net/browse/ENT-3896

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [x] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [x] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [x] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: enterprise/models.py [hunk 1]
```diff
@@ -708,7 +708,8 @@ def get_link_by_email(self, user_email):
             pass
 
         try:
-            return PendingEnterpriseCustomerUser.objects.get(user_email=user_email)
+            # return the first element in case of admin/learner with multiple pending enterprise associations.
+            return PendingEnterpriseCustomerUser.objects.filter(user_email=user_email).first()
```
Comment (adamstankiewicz) [inline enterprise/models.py:712]: This is OK to return just the first element if there are multiple. This function is used downstream to determine if the user has any existing pending links to any customer.

File: tests/test_enterprise/api/test_views.py [hunk 2]
```diff
@@ -431,7 +431,7 @@ def test_post_enterprise_customer_user_logged_out(self):
 
     @ddt.data(
         {'is_staff': True, 'user_exists': True, 'ecu_exists': True, 'pending_ecu_exists': False, 'status_code': 204},
-        {'is_staff': True, 'user_exists': True, 'ecu_exists': False, 'pending_ecu_exists': True, 'status_code': 204},
+        {'is_staff': True, 'user_exists': True, 'ecu_exists': False, 'pending_ecu_exists': True, 'status_code': 201},
```
Comment (adamstankiewicz) [inline tests/test_enterprise/api/test_views.py:434]: As the view is implemented, if we post to the pending enterprise customer user API for a user that already exists and has a PendingEnterpriseCustomerUser record, we opt to link the user by creating an `EnterpriseCustomerUser` record for that user/customer.

As such, we would expect a 201 in this case, not a 204.

File: enterprise/models.py (hunk 1 above)
Comment (edx-abolger) [reply enterprise/models.py:712]: Then perhaps make it a boolean?

Comment (adamstankiewicz) [reply enterprise/models.py:712]: We do expect it to the function to return either a `EnterpriseCustomerUser` or `PendingEnterpriseCustomerUser` object as we refer to its `enterprise_customer` relationship, so returning a boolean here wouldn't quite make sense.

File: enterprise/signals.py [hunk 3]
```diff
@@ -63,27 +64,21 @@ def handle_user_post_save(sender, **kwargs):  # pylint: disable=unused-argument
     if user_instance is None:
         return  # should never happen, but better safe than 500 error
 
-    try:
-        pending_ecu = PendingEnterpriseCustomerUser.objects.get(user_email=user_instance.email)
-    except PendingEnterpriseCustomerUser.DoesNotExist:
-        pending_ecu = None
+    pending_ecus = PendingEnterpriseCustomerUser.objects.filter(user_email=user_instance.email)
```
Comment (bbaker6225) [inline enterprise/signals.py:67]: If we’re only looking up by email, no need to index together with customer. I misunderstood that we’d be using both in the lookup

**Review by bbaker6225 — APPROVED** [review]
One small comment, but otherwise lgtm

File: enterprise/signals.py (hunk 3 above)
Comment (adamstankiewicz) [reply enterprise/signals.py:67]: Most other places where we call `PendingEnterpriseCustomerUser.objects.get` already supplied both a `user_email` and an `enterprise_customer`. Would it make sense to keep the index together with customer for those cases?

Comment (bbaker6225) [reply enterprise/signals.py:67]: Yep that makes sense. You can also add user email on it’s own for this case like in the docs example:

class Meta:
        indexes = [
            models.Index(fields=['last_name', 'first_name']),
            models.Index(fields=['first_name'], name='first_name_idx'),
        ]

Comment (adamstankiewicz) [reply enterprise/signals.py:67]: Ah, got it. Cool, will update!

**Pull request merged**